python/lsst/ts/tcpip/client.py

```diff
@@ -0,0 +1,206 @@
+    """A TCP/IP socket client.
+
+    A thin wrapper around `asyncio.open_connection`. Like that function,
+    it can only be used to connect once. Construct a new instance each
+    time you wish to make a new connection.
+
+    Parameters
+    ----------
+    host : `str` | `None`
+        IP address.
+    port : `int` | `None`
+        IP port.
+    log : `logging.Logger`
+        Logger.
+    connect_callback : callable or `None`, optional
+        Asynchronous function to call when the connection state changes.
+        If the other end (server) closes the connection, it may take
+        ``monitor_connection_interval`` seconds or longer to notice.
+        The function receives one argument: this `Client`.
+    monitor_connection_interval : `float`, optional
+        Interval between checking if the connection is still alive (seconds).
+        Defaults to DEFAULT_MONITOR_CONNECTION_INTERVAL.
+        If ≤ 0 then do not monitor the connection at all.
+        Monitoring is only useful if you do not regularly read from the reader
+        using the read methods of this class (or copying what they do
+        to detect and report hangups).
+    name : `str`
+        Optional name used for log messages.
+    **kwargs : `dict` [`str`, `typing.Any`]
+        Additional keyword arguments for `asyncio.open_connection`,
+        beyond host and port.
+
+    Attributes
+    ----------
+    host : `str` | `None`
+        IP address; the ``host`` constructor argument.
+    port : `int` | `None`
+        IP port; the ``port`` constructor argument.
+    log : `logging.Logger`
+        A child of the ``log`` constructor argument.
+    name : `str`
+        The ``name`` constructor argument.
+    reader : `asyncio.StreamReader` or None
+        Stream reader to read data from the server.
+        This will be a stream reader (not None) if `connected` is True.
+    writer : `asyncio.StreamWriter` or None
+        Stream writer to write data to the server.
+        This will be a stream writer (not None) if `connected` is True.
+    should_be_connected : `bool`
+        True if the connection was made and close not called.
+        The connection was unexpectedly lost if ``should_be_connected``
+        is true and ``connected`` is false (unless you close the connection
+        by calling `basic_close` or manually closing ``writer``).
+    start_task : `asyncio.Future`
+        Future that is set done when the connection is made.
+    done_task : `asyncio.Future`
+        Future that is set done when this client is closed, at which point
+        it is no longer usable.
+
+    Raises
+    ------
+    `TypeError`
+        If `connect_callback` is synchronous.
+
+    Notes
+    -----
+    See the User Guide for an example.
+
+    Always wait for `start_task` after constructing an instance,
+    before using the instance. This indicates the client has connected.
+
+    Always check that `connected` is True before reading or writing.
```

Comment:
Should I check "should_be_connected" as well? This is why I am very confusing as a user.

Reply:
Yes. I have overhauled the documentation (and removed this line, because the read and write methods handle this).